pkgs/development/compilers/openjdk/generic.nix

```diff
@@ -0,0 +1,664 @@
+        sha256 = "082lmc30x64x583vqq00c8y0wqih3y4r0mp1c4bqq36l22qv6b6r";
+      })
+    ]
+    ++ lib.optionals (featureVersion == "17") [
+      # Patch borrowed from Alpine to fix build errors with musl libc and recent gcc.
+      # This is applied anywhere to prevent patchrot.
+      (fetchurl {
+        url = "https://git.alpinelinux.org/aports/plain/community/openjdk17/FixNullPtrCast.patch?id=41e78a067953e0b13d062d632bae6c4f8028d91c";
+        sha256 = "sha256-LzmSew51+DyqqGyyMw2fbXeBluCiCYsS1nCjt9hX6zo=";
+      })
+    ]
+    ++ lib.optionals atLeast11 [
+      # Fix build for gnumake-4.4.1:
+      #   https://github.com/openjdk/jdk/pull/12992
+      (fetchpatch {
+        name = "gnumake-4.4.1";
+        url = "https://github.com/openjdk/jdk/commit/9341d135b855cc208d48e47d30cd90aafa354c36.patch";
+        hash = "sha256-Qcm3ZmGCOYLZcskNjj7DYR85R4v07vYvvavrVOYL8vg=";
+      })
+    ]
+    ++ lib.optionals (featureVersion == "17") [
+      # Backport fixes for musl 1.2.4 which are already applied in jdk21+
+      # Fetching patch from chimera because they already went through the effort of rebasing it onto jdk17
+      (fetchurl {
+        name = "lfs64.patch";
+        url = "https://raw.githubusercontent.com/chimera-linux/cports/4614075d19e9c9636f3f7e476687247f63330a35/contrib/openjdk17/patches/lfs64.patch";
+        hash = "sha256-t2mRbdEiumBAbIAC0zsJNwCn59WYWHsnRtuOSL6bWB4=";
+      })
+    ]
+    ++ lib.optionals (!headless && enableGtk) [
+      (
+        if atLeast17 then
+          ./17/patches/swing-use-gtk-jdk13.patch
+        else if atLeast11 then
+          ./11/patches/swing-use-gtk-jdk10.patch
+        else
+          ./8/patches/swing-use-gtk-jdk8.patch
+      )
+    ];
+
+  nativeBuildInputs =
+    [
+      pkg-config
+    ]
+    ++ lib.optionals atLeast11 [
+      autoconf
+    ]
+    ++ lib.optionals (!atLeast11) [
+      lndir
+    ]
+    ++ [
+      unzip
+    ]
+    ++ lib.optionals atLeast21 [
+      ensureNewerSourcesForZipFilesHook
+    ];
+
+  buildInputs =
+    [
+      # TODO: Many of these should likely be in `nativeBuildInputs`.
+      cpio
+      file
+      which
+      zip
+      perl
+      zlib
+      cups
+      freetype
+    ]
+    ++ lib.optionals (atLeast11 && !atLeast21) [
+      harfbuzz
+    ]
+    ++ [
+      alsa-lib
+      libjpeg
+      giflib
+    ]
+    ++ lib.optionals atLeast11 [
+      libpng
+      zlib # duplicate
+      lcms2
+    ]
+    ++ [
+      libX11
+      libICE
+    ]
+    ++ lib.optionals (!atLeast11) [
+      libXext
+    ]
+    ++ [
+      libXrender
+    ]
+    ++ lib.optionals atLeast11 [
+      libXext
+    ]
+    ++ [
+      libXtst
+      libXt
+      libXtst # duplicate
+      libXi
+      libXinerama
+      libXcursor
+      libXrandr
+      fontconfig
+      jdk-bootstrap'
+    ]
+    ++ lib.optionals (!headless && enableGtk) [
+      (if atLeast11 then gtk3 else gtk2)
+      glib
+    ];
```

Comment:
They do, because these lists factor in to the final `.drv` file. That’s why I’ve marked the duplicates and maintained the version‐conditional order.

Reply:
BTW, the ordering is semantically meaningful when multiple packages contain the same libraries or similar.

Reply:
There can definitely be semantic meaning, but it feels messier with it as split up as it is. Given it causes rebuilds this can be marked resolved and thought about later.

Reply:
Yeah, I will definitely try to clean this stuff up for 25.05. It will need adjusting for Darwin anyway.